[CSM-527] Add tests on ungrouped inputs picking in tx creation

hithroc opened the pull request:
Added tests for both grouped and ungrouped input selection policies.

Each transaction creation test is now tested with both grouped and ungrouped selection policy.

Note: this PR directly depends on #1826

Martoon-00 commented on node/test/Test/Pos/Client/Txp/UtilSpec.hs:
```diff
@@ -257,6 +263,30 @@ feeForManyAddressesSpec inputSelectionPolicy manyAddrs =
+        Left err -> stopProperty $ pretty err
+        Right (txAux, _) ->
+            let picked = length . _txInputs . taTx $ txAux
+            in unless (picked == (1 :: Int)) . stopProperty
```
Is `:: Int ` needed here? `picked` seem to already be `Int`

hithroc replied:
Nope, it is a leftover after me testing stuff.

Martoon-00 commented on node/test/Test/Pos/Client/Txp/UtilSpec.hs:
```diff
@@ -51,17 +52,32 @@ spec = giveGtConf $ giveNodeConf $ giveInfraConf $ giveUpdateConf $ giveCoreConf
-    prop manyAddressesTo1Desc manyAddressesTo1Spec
-    prop manyAddressesToManyDesc manyAddressesToManySpec
+    let inputSelectionPolicies =
+            [("Grouped inputs", OptimizeForSecurity)
```
Please add a space after `[` and `,`

Martoon-00 commented on node/test/Test/Pos/Client/Txp/UtilSpec.hs:
```diff
@@ -51,17 +52,32 @@ spec = giveGtConf $ giveNodeConf $ giveInfraConf $ giveUpdateConf $ giveCoreConf
+    sequence_ $ do
+        (funcDesc, TestFunctionWrapper func) <- testSpecs
+        (inputSelectionDesc, inputSelectionPolicy) <- inputSelectionPolicies
+        return $ prop (inputSelectionDesc <> ": " <> funcDesc) (func inputSelectionPolicy)
```
I think it's better to do `prop inputSelectionDesc . prop funcDesc` to get true hierarchy

Martoon-00 commented on node/test/Test/Pos/Client/Txp/UtilSpec.hs:
```diff
@@ -51,17 +52,32 @@ spec = giveGtConf $ giveNodeConf $ giveInfraConf $ giveUpdateConf $ giveCoreConf
+-- GHC doesn't support impredicative polymorphism so we need a wrapper
+-- for the list below to typecheck.
+data TestFunctionWrapper
+    = forall prop. (Testable prop) => TestFunctionWrapper (InputSelectionPolicy -> prop)
```
This is neat, but you put only those properties to that list which can depend on `InputSelectionPolicy`, so essentially anything except `TxpTestProperty` hardly can be used there

hithroc replied:
I wouldn't go for it, but there was one function that required it:

    feeForManyAddressesSpec :: InputSelectionPolicy -> Bool -> TxpTestProperty ()

Martoon-00 replied:
Oh harsh, okay

hithroc commented:
I've adressed all the comments.

The styling is fixed, redundant type signatures removed, and the tests now use HSpec hierarchy to group them by input selection policy.

Martoon-00 commented on node/test/Test/Pos/Client/Txp/UtilSpec.hs:
```diff
@@ -73,10 +73,11 @@ createMTxSpec = do
-        return $ prop (inputSelectionDesc <> ": " <> funcDesc) (func inputSelectionPolicy)
+
+    sequence_ $ inputSelectionPolicies <&> \(inputSelectionDesc, policy) ->
+        describe inputSelectionDesc . sequence_ $
```
now `traverse_` can be used here

Martoon-00 commented:
LGTM, let's wait for #1826 to be merged

Martoon-00 commented on node/test/Test/Pos/Client/Txp/UtilSpec.hs:
```diff
@@ -51,17 +52,35 @@ spec = giveGtConf $ giveNodeConf $ giveInfraConf $ giveUpdateConf $ giveCoreConf
+
+    for_ inputSelectionPolicies $ \(inputSelectionDesc, policy) ->
+        describe inputSelectionDesc . sequence_ $
+            map (\(funcDesc, TestFunctionWrapper func) -> prop funcDesc (func policy)) testSpecs
```
And one another here :smile:

dcoutts commented:
Tests are always good. Once #1826 is merged, can I ask for this one also to be rebased to squash the minor fixups, review fixes etc.

hithroc commented:
Now that #1826 is merged, I've squashed all the commits and rebased them on top of cardano-sl-1.0 branch (because the original CSL-1793 branch was rebased).

Martoon-00 commented:
When it gets merged, if you have time please proceed with merging `cardano-sl-1.0` to `master`.
Functions like `makeManyAddressesToManyParams` were modified in `master` so you may have many conflicts and probably will need to revise your changes a bit
